src/main/java/org/microg/nlp/backend/nominatim/BackendService.java

```diff
@@ -304,4 +408,163 @@ String retrieveString() {
+        String[] projection = {
+            LocationAddressCache.COLUMN_NAME_ADDRESS
+        };
+
+        double latitudeLow = latitude - 0.0001;
+        double latitudeHigh = latitude + 0.0001;
+        double longitudeLow = longitude - 0.0001;
+        double longitudeHigh = longitude + 0.0001;
+
+        String selection = LocationAddressCache.COLUMN_NAME_LONGITUDE + " <= ? and " +
+                           LocationAddressCache.COLUMN_NAME_LONGITUDE + " >= ? and " +
+                           LocationAddressCache.COLUMN_NAME_LATITUDE + " <= ? and " +
+                           LocationAddressCache.COLUMN_NAME_LATITUDE + " >= ? and " +
+                           LocationAddressCache.COLUMN_NAME_LOCALE + " = ? ";
+        String[] selectionArgs = { String.valueOf(longitudeHigh),
+                                   String.valueOf(longitudeLow),
+                                   String.valueOf(latitudeHigh),
+                                   String.valueOf(latitudeLow),
+                                   locale };
+
+        Cursor cursor = db.query(
+            LocationAddressCache.TABLE_NAME,
+            projection,
+            selection,
+            selectionArgs,
+            null,
+            null,
+            null
+            );
+
+        if (!cursor.moveToNext()) {
+            cursor.close();
+            return null;
+        }
+
+        byte[] cachedAddressBytes = cursor.getBlob(
+                                      cursor.getColumnIndexOrThrow(LocationAddressCache.COLUMN_NAME_ADDRESS));
+        cursor.close();
+
+        return ReverseGeocodingCacheDbHelper.getAddressFromBytes(cachedAddressBytes);
+    }
+
+    private boolean recordDateIsNotValidOrIsTooOld(String recordCreatedTxt) {
+        Calendar now = Calendar.getInstance();
+        Calendar calendarRecordCreated = Calendar.getInstance();
+        try {
+            Date recordCreated = iso8601Format.parse(recordCreatedTxt);
+            calendarRecordCreated.setTime(recordCreated);
+        } catch (ParseException ex) {
+            Log.e(TAG, "Date of date created could not be parsed. Orignal value is " + recordCreatedTxt, ex);
+            appendLog(TAG, ex.getMessage(), ex);
+            return true;
+        }
+
+        int timeToLiveRecordsInCacheInHours = Integer.parseInt(
+                        PreferenceManager.getDefaultSharedPreferences(this).getString(SettingsActivity.LOCATION_CACHE_LASTING_HOURS, "720"));
+
+        calendarRecordCreated.add(Calendar.HOUR_OF_DAY, timeToLiveRecordsInCacheInHours);
+        return calendarRecordCreated.before(now);
+    }
+
+    private void deleteRecordFromTable(Integer recordId) {
+        SQLiteDatabase db = mDbHelper.getWritableDatabase();
+        String selection = LocationAddressCache._ID + " = ?";
+        String[] selectionArgs = { recordId.toString() };
+        db.delete(LocationAddressCache.TABLE_NAME, selection, selectionArgs);
+    }
+
+    private byte[] getAddressAsBytes(Address address) {
```

Comment:
Would be better in ReverseGeocodingCacheDbHelper next to getAddressFromBytes() (or: put both into ReverseGeocodingCacheContract)

Reply:
fixed